server/qrops.go

```diff
@@ -0,0 +1,273 @@
+
+	if !ok || len(allowProfileImage) == 0 {
+		// we default to false when this flag was not provided
+		// so someone does not want to allowProfileImage on their QR Image
+		// fine then :)
+		return false
+	}
+
+	LogoOnImage, err := strconv.ParseBool(allowProfileImage[0])
+
+	if err != nil {
+		// maybe for fun someone tries to send non-boolean values to this flag
+		// we also default to false in that case
+		return false
+	}
+
+	// if we reach here its most probably true, but we pass the flag back
+	return LogoOnImage
+
+}
+
+func getLevel(params url.Values, logger *zap.Logger) string {
+	level, ok := params["level"]
+
+	if !ok || len(level) == 0 {
+		// we default to MediumLevel of error correction when the level flag
+		// is not passed.
+		return "2"
+	}
+
+	return level[0]
+}
+
+func getKeyUID(params url.Values, allowProfileImages bool, logger *zap.Logger) string {
+	keyUID, ok := params["keyUid"]
+
+	// if allowProfileImage was also not passed
+	// in that case we don't care and can return an empty string back
+	if !allowProfileImages {
+		return ""
+	}
+
+	// the keyUID was not passed, which is a requirement to get the multiaccount image,
+	// so we log this error
+	if !ok || len(keyUID) == 0 {
+		logger.Error("[qrops-error] A keyUID is required to put logo on image and it was not passed in the parameters")
+	}
+
+	return keyUID[0]
+}
+
+func getImageName(params url.Values, allowProfileImages bool, logger *zap.Logger) string {
+	imageName, ok := params["imageName"]
+
+	// if allowProfileImage was also not passed
+	// in that case we don't care and can return an empty string back
+	if !allowProfileImages {
+		return ""
+	}
+
+	//if the imageName was not passed, we default to "thumbnail"
+	if !ok || len(imageName) == 0 {
+		return "thumbnail"
+	}
+
+	return imageName[0]
+}
+
+func getSize(params url.Values, logger *zap.Logger) int {
+	size, ok := params["size"]
+
+	//if the size was not passed, we return nil;
+	if !ok || len(size) == 0 {
+		return 0
+	}
+	imageToBeResized, err := strconv.Atoi(size[0])
+
+	if err != nil {
+		logger.Error("[qrops-error] could not convert string to int for size param ", zap.Error(err))
+		return 0
+	}
+
+	return imageToBeResized
+}
+
+func generateQRBytes(params url.Values, logger *zap.Logger, multiaccountsDB *multiaccounts.Database) []byte {
+
+	qrURLBase64Decoded := getQrURL(params, logger)
+	superImposeLogoOnQR := getAllowProfileImage(params, logger)
+	level := getLevel(params, logger)
+	correctionLevel := qrcode.ErrorCorrectionMedium
+	keyUID := getKeyUID(params, superImposeLogoOnQR, logger)
+	imageName := getImageName(params, superImposeLogoOnQR, logger)
+	size := getSize(params, logger)
+
+	switch level {
+	case "1":
+		correctionLevel = qrcode.ErrorCorrectionLow
+	case "2":
+		correctionLevel = qrcode.ErrorCorrectionMedium
+	case "3":
+		correctionLevel = qrcode.ErrorCorrectionQuart
+	case "4":
+		correctionLevel = qrcode.ErrorCorrectionHighest
+	}
```

Comment:
This code is basically building a configuration. Perhaps these values could be encapsulated in a struct.

Additionally each of these functions consume the same variables, `params` and `logger`. Is it possible that this functionality could belong to a single struct that manages holds these variables?

Something like:

```go
type qrConfig struct {
	Data            string
	WithLogo        bool
	CorrectionLevel qrcode.ErrorCorrectionLevel
	KeyUID          string
	ImageName       string
	Size            uint
}
``` 

Then your consuming code can handle the struct fields and you get default empty values for free (`""`, `0`, etc)

Reply:
now that I'm more confident, I shall attempt this.

Reply:
fixed here : https://github.com/status-im/status-go/pull/3172/commits/3f45a500e394591b67c7dbedc0a07ac3b0ae1623